apps/mobile/src/features/notifications/use-notifications.ts

```diff
@@ -0,0 +1,167 @@
+  const releasePushNotifications = useReleasePushNotificationsFlag();
+
+  const registerNotificationsBTC = useCallback(async () => {
+    const notificationToken = await messaging().getToken();
+    await registerToken({ addresses: bitcoinAddresses, notificationToken, chain: 'bitcoin' });
+  }, [registerToken, bitcoinAddresses]);
+
+  const registerNotificationsSTX = useCallback(async () => {
+    const notificationToken = await messaging().getToken();
+
+    await registerToken({ addresses: stacksAddresses, notificationToken, chain: 'stacks' });
+  }, [registerToken, stacksAddresses]);
+
+  const registerInAppNotifications = useCallback(() => {
+    messaging().onMessage(remoteMessage => {
+      if (remoteMessage.notification?.body) {
+        displayToast({ title: remoteMessage.notification.body, type: 'success' });
+      }
+    });
+  }, [displayToast]);
+
+  const unregisterNotifications = useCallback(async () => {
+    if (releasePushNotifications) {
+      // Delete FCM token from Firebase
+      await messaging().deleteToken();
+    }
+  }, [releasePushNotifications]);
+
+  const requestUserPermissions = useCallback(async () => {
+    if (Platform.OS === 'ios') {
+      const { status: existingStatus } = await Notifications.getPermissionsAsync();
+      let finalStatus = existingStatus;
+      if (existingStatus !== 'granted') {
+        const { status } = await Notifications.requestPermissionsAsync();
+        finalStatus = status;
+      }
+      // TODO: analytics?
+      if (finalStatus !== 'granted') {
+        return;
+      }
+    }
+
+    if (Platform.OS === 'android') {
+      await PermissionsAndroid.request('android.permission.POST_NOTIFICATIONS');
+    }
+  }, []);
+
+  const registerNotifications = useCallback(async () => {
+    if (releasePushNotifications) {
+      await registerNotificationsBTC();
+      await registerNotificationsSTX();
+    }
+  }, [registerNotificationsBTC, registerNotificationsSTX, releasePushNotifications]);
+
+  return {
+    registerNotificationsBTC,
+    registerNotificationsSTX,
+    registerNotifications,
+    unregisterNotifications,
+    requestUserPermissions,
+    registerInAppNotifications,
+  };
+}
+
+export function useWatchNotificationAddresses() {
+  const btcAddresses = useBitcoinAddresses();
+  const stacksAddresses = useStacksSignerAddresses();
+  const releasePushNotifications = useReleasePushNotificationsFlag();
+  const { notificationsPreference } = useSettings();
+  const {
+    requestUserPermissions,
+    registerNotificationsBTC,
+    registerNotificationsSTX,
+    registerInAppNotifications,
+    unregisterNotifications,
+  } = useRegisterNotifications();
+
+  useEffect(() => {
+    if (releasePushNotifications && notificationsPreference === 'enabled') {
+      void requestUserPermissions();
+      registerInAppNotifications();
+    }
+  }, [
+    releasePushNotifications,
+    notificationsPreference,
+    registerInAppNotifications,
+    requestUserPermissions,
+  ]);
```

Comment:
Perhaps this is just the way to do it, but I get scared seeing all these render-time side effects. Code like this is common in react but it runs this every render? 🤯